### ISSUE TITLE
How to add trait in tripal content types

### ISSUE THREAD

**Issue by cornellyujy** [UofS-Pulse-Binfo/analyzedphenotypes#133]
Hi @laceysanderson , in our tripal, we have no trait in tripal content types section. I have to add trait in tripal content types.
Here are my steps, and I am not if make sense.
#step1. select cvterm as the trait table
![image](https://user-images.githubusercontent.com/59180297/162232457-f1159129-0399-48e8-a0a4-0b71da6bcfe2.png)
#step2. select All records that belong to a single controlled vocabulary
![image](https://user-images.githubusercontent.com/59180297/162232638-d2cc020f-7924-4282-901f-266c6fcfa706.png)
#step3. select a controlled vocabulary
![image](https://user-images.githubusercontent.com/59180297/162233121-884b96b9-1035-476e-84e4-21a3d6f6c64a.png)
#step4. create content types
![image](https://user-images.githubusercontent.com/59180297/162236624-54282581-515c-4e67-addd-6c22dfe8d167.png)

For the step3, I am not sure which controlled vocabulary should be selected. You know last time you told me that the 3 CVs (traits, methods, and Units) should have one db, but I have created two types: 1 DB for the 3 CVs, and 3 DBs for 3 CVs. Now, if we cancel the 3 to 3, which db should be added. Thanks~

**Comment by laceysanderson** [comment]
Hello, those steps are right. As mentioned in the previous issue, I made a slight mistake and the 1DB for the 3 CVs is the correct approach. For the trait content type you will want to choose the CV you select for the traits to go in. The example used in the docs is “Phenotypic Traits: Tripalus” and your screenshot looks correct. 

To double-check, go to the set-up ontologies page and check that the "Trait" CV selected (the top dropdown in that column; circled green in the following screenshot) matches the cv you selected for the Trait content type.
![setup 1 traitontologies](https://user-images.githubusercontent.com/1566301/163065632-3f2cec0f-65ba-427e-8e26-41052b640c83.png)

**Comment by cornellyujy** [comment]
Thanks Lacey.
Yes. I created the trait content type in 'Tripal Contents Types' following above steps. When running the 'sudo drush trp-run-jobs' command, I got this error:
![image](https://user-images.githubusercontent.com/59180297/164306242-9be24468-18e2-41b4-99a9-857152c7ad0f.png)
So, is there anything wrong when creating the 3 CVs in Admin > Tripal > Data Loaders > Chado Vocabularies > Manage Chado CVs? 

I checked the steps for creating 3CVs, and the below figs are the panels for creating CVs
![image](https://user-images.githubusercontent.com/59180297/164309320-82545fe8-3d1f-44f0-91a8-b389ac54cad5.png)
There is no information input.
After creation, they have default information for the 3CVs:
![image](https://user-images.githubusercontent.com/59180297/164309807-e28eed88-be63-41cc-86c4-714ad0a81e6c.png)
![image](https://user-images.githubusercontent.com/59180297/164309898-5bd7ef7f-bd3a-4d35-8d96-495eeae7ca43.png)
![image](https://user-images.githubusercontent.com/59180297/164310013-dae3e469-afff-4942-a513-1a0eeeb12a69.png)
And I noticed that the accessions were inputted into chado.dbxref table. I am not sure why it have default information:
![image](https://user-images.githubusercontent.com/59180297/164313376-9666c120-d91e-4f98-b2aa-e3fa00feafea.png)
Should I update the CVs first? Here is the trait CV.
![image](https://user-images.githubusercontent.com/59180297/164313828-6355dec3-c252-4cd5-b001-60db0dce01e0.png)
How to fix the 'term name' and 'accession'. Thanks~

**Comment by laceysanderson** [comment]
Hi @cornellyujy,

Based on the screenshots above 
- your Trait Content Type is created correctly ✅ 
- your CVs/DBs are also correct ✅ 
- You do not want to change the terms you have in your CV as that can cause problems with the traits. You likely have those terms available because you imported one of my test files 🙂 

The error you see when creating the content type:
- would not have impacted the content type creation 🎉 
- it's actually related to adding fields to the created content type
- It's saying you do not have the UO:unit term in your site. This term is used in our hp__phenotypic_variability and sio__study fields and it should have been added automatically when you set the terms for the module 🤔 

Solution:
1. On the command-line within your Tripal site run the following command to add any missing default terms: `drush php-eval "ap_load_defaultontology();"`
2. Now navigate to the "Manage Fields" page for you Trait content type by clicking on "Structure" and then "Tripal Content Types" in your admin toolbar at the top of the page, then on the "manage fields" link beside the Trait Content Type.
3. Finally click "Check for new fields" at the top of the page.

That should ensure you have all the fields added to your Trait Content Type and they will now have all the terms they need.

To play with Trait pages further you will want to "Publish" them so you can see the example content :-)

**Comment by cornellyujy** [comment]
Hi @laceysanderson , Thanks for your response. Now I have loaded the phenotypic data into chado table.
You mean I should 'Publish' the trait pages. How to publish? I tried to publish, but failed.
![image](https://user-images.githubusercontent.com/59180297/164529645-5fba35f7-bae2-435e-8414-fc1f4b234fc4.png)
Here is the drush command:
![image](https://user-images.githubusercontent.com/59180297/164529946-e89034f5-0695-4e2f-99aa-8b6f49881b5a.png)
Thanks~

**Comment by laceysanderson** [comment]
I have never seen that before nor can I duplicate it on a fresh Tripal site. I am so confused as that error is implying that the vocabulary attached to that content type no longer exists... but your screenshots show it does. 

Can you double-check that the vocabulary shown in the Tripal Content edit page for "Trait" is still available in your Chado database? Also, look in your pgAdmin at the `chado_bundle` table. There should be an entry in there where the bundle_id = that of your tripal content type (i.e. bio_data_[this is the bundle_id] ). The type_value of that content type should be the same as the cv_id for your "Phenotypic Traits: Cucurbita" CV.

**Comment by cornellyujy** [comment]
Thanks Lacey. I got it. I deleted the 'Phenotypic Traits: Cucurbita' after creating Trait content type. And recreated the Phenotypic Traits: Cucurbita, so the Trait can not get the new cv_id. I wondered that 'Phenotypic Traits: Cucurbita' was the only one phenotypic traits of Cucurbita and CuCurbita is one genus. So if I add new genus, should I create a new trait content type for the newly added genus? I am not sure if I am saying this clearly. 
I want to know if we can create a trait content type for all genus, not only for one genus.
![image](https://user-images.githubusercontent.com/59180297/165326765-afde408c-8324-4b87-b48e-6c263102ed72.png)
For this example, it depend on the cv of only one genus. so this trait is a little hard to imply other genus. Or is there anything wrong when creating the trait.
Thanks~

**Comment by laceysanderson** [comment]
>  I deleted the 'Phenotypic Traits: Cucurbita' after creating Trait content type. And recreated the Phenotypic Traits: Cucurbita, so the Trait can not get the new cv_id.

Exactly. Everything in Tripal and my modules use the keys so creating new will always break the connection.

>  I wondered that 'Phenotypic Traits: Cucurbita' was the only one phenotypic traits of Cucurbita and CuCurbita is one genus. So if I add new genus, should I create a new trait content type for the newly added genus?

Yes, there will be one "Trait" Content Type per genus. For example, in KnowPulse we have a "Lentil Traits" for Lens genus (with multiple species) and a "Common Bean Traits" for Phaseolus genus (again with multiple species).
